llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp

```diff
@@ -52,12 +52,60 @@ class RISCVCodeGenPrepare : public FunctionPass,
   }
 
   bool visitInstruction(Instruction &I) { return false; }
+  bool visitBinaryOperator(BinaryOperator &BO);
   bool visitAnd(BinaryOperator &BO);
   bool visitIntrinsicInst(IntrinsicInst &I);
 };
 
 } // end anonymous namespace
 
+/// InstCombine will canonicalize selects of binary ops where the identity is
+/// zero to zexts:
+///
+/// select c, (add x, 1), x -> add x, (zext c)
+///
+/// On RISC-V though, a zext of an i1 vector will be lowered as a vmv.v.i and a
+/// vmerge.vim:
+///
+///       vmv.v.i v12, 0
+///       vmerge.vim      v9, v12, 1, v0
+///       vadd.vv v8, v8, v9
+///
+/// Reverse this transform so that we pull the select outside of the binary op,
+/// which allows us to fold it into a masked op:
+///
+///       vadd.vi v8, v8, 1, v0.t
+bool RISCVCodeGenPrepare::visitBinaryOperator(BinaryOperator &BO) {
+  if (!BO.getType()->isVectorTy())
+    return false;
+
+  // TODO: We could allow sub if we did a non-commutative match
+  Constant *Identity = ConstantExpr::getIdentity(&BO, BO.getType());
+  if (!Identity || !Identity->isNullValue())
+    return false;
+
+  using namespace PatternMatch;
+
+  Value *Mask, *RHS;
+  if (!match(&BO, m_c_BinOp(m_OneUse(m_ZExt(m_Value(Mask))), m_Value(RHS))))
+    return false;
+
+  if (!Mask->getType()->isIntOrIntVectorTy(1))
+    return false;
+
+  IRBuilder<> Builder(&BO);
+  Value *Splat = ConstantInt::get(BO.getType(), 1);
+  Value *NewBO = Builder.CreateBinOp(BO.getOpcode(), RHS, Splat);
```

Comment:
This requires a freeze on RHS

Reply:
I think I'm missing something here. We aren't increasing the number of uses of RHS and alive doesn't seem to complain. And we're not speculatively executing anything?

Reply:
Isn't `RHS` used by `NewBO` and `Select`?

Reply:
Sorry yes, I missed that. 

I'm guessing the reason why alive doesn't require the freeze then is because `(add (zext c), undef) ~> undef`. So `select c, (add undef, 1), undef` would also be just as undefined?

Reply:
Freeze is unnecessary here because we only use one arm of the select inst.

Reply:
Why does `foldBinOpIntoSelect` use a freeze then?

Reply:
> Why does `foldBinOpIntoSelect` use a freeze then?

https://github.com/llvm/llvm-project/blob/8bb9443333e0117ab61feecce9de339b11b924fc/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2375-L2388

IIUC the freeze can be eliminated as well.

Reply:
Related patch: https://reviews.llvm.org/D118644